bikeshed/InputSource.py

```diff
@@ -0,0 +1,155 @@
+
+    @abstractmethod
+    def __str__(self) -> str:
+        pass
+
+    @abstractmethod
+    def read(self) -> InputContent:
+        """Fully reads the source.
+        """
+        pass
+
+    def hasDirectory(self) -> bool:
+        """Only some InputSources have a directory."""
+        return False
+
+    def directory(self) -> str:
+        """Suitable for passing to subprocess(cwd=)."""
+        raise TypeError("{} instances don't have directories.".format(type(self)))
+
+    def relative(self, relativePath) -> Optional[InputSource]:
+        """Resolves relativePath relative to this InputSource.
+
+        For example, InputSource("/foo/bar/baz.txt").relative("quux/fuzzy.txt")
+        will be InputSource("/foo/bar/quux/fuzzy.txt").
+
+        If this source type can't find others relative to itself, returns None.
+        """
+        return None
+
+    def mtime(self) -> Optional[float]:
+        """Returns the last modification time of this source, if that's known.
+        """
+        return None
+
+
+class StdinInputSource(InputSource):
+    def __init__(self, sourceName: str):
+        assert sourceName == "-"
+        self.type = "stdin"
+        self.sourceName = sourceName
+        self.content = None
+
+    def __str__(self) -> str:
+        return "-"
+
+    def read(self) -> InputContent:
+        return InputContent(sys.stdin.readlines(), None)
+
+
+class UrlInputSource(InputSource):
+    def __init__(self, sourceName: str):
+        assert sourceName.startswith("https:")
+        self.sourceName = sourceName
+        self.type = "url"
+
+    def __str__(self) -> str:
+        return self.sourceName
+
+    def read(self) -> InputContent:
+        response = requests.get(self.sourceName)
+        if response.status_code != 200:
+            # This matches the OSErrors expected by older uses of FileInputSource.
+            raise FileNotFoundError(errno.ENOENT, response.text, self.sourceName)
+        date = None
+        if "Date" in response.headers:
+            # Use the response's Date header, although servers don't always set
+            # this according to the last change to the file.
+            date = email.utils.parsedate_to_datetime(response.headers["Date"]).date()
+        return InputContent(response.text.splitlines(True), date)
+
+    def relative(self, relativePath) -> UrlInputSource:
+        return UrlInputSource(urllib.parse.urljoin(self.sourceName, relativePath))
+
+
+class FileInputSource(InputSource):
+    def __init__(self, sourceName: str):
+        self.sourceName = sourceName
+        self.type = "file"
+        self.content = None
+
+    def __str__(self) -> str:
+        return self.sourceName
+
+    def read(self) -> InputContent:
+        with io.open(self.sourceName, "r", encoding="utf-8") as f:
+            return InputContent(
+                f.readlines(),
+                datetime.fromtimestamp(os.path.getmtime(self.sourceName)).date(),
+            )
```

Comment:
I formatted this file with [black](https://black.readthedocs.io/en/stable/), but it might have different style opinions than you do... Which formatter do you use?

Reply:
I don't! Probably should at some point, but for now it's hand-formatted. I'm fine with these decisions.